File: tui/internal/crypto/beacon_test.go
package crypto

import (
	"bytes"
	"encoding/base64"
	"encoding/hex"
	"os"
	"testing"

	"golang.org/x/crypto/chacha20poly1305"
)

// knownRoomKey is the fixed room key used by the pinned beacon vectors.
func knownRoomKey() RoomKey {
	var rk RoomKey
	for i := range rk.Key {
		rk.Key[i] = byte(i) // 0x00..0x1f, fixed
	}
	return rk
}

// TestBeaconKeyVector pins the exact beacon-key derivation for a known room key
// (§1.2), the same way protocol/signing_test pins its bytes. The web reader derives
// the identical key value; if either side changes, beacon links stop decrypting.
func TestBeaconKeyVector(t *testing.T) {
	bk := BeaconKey(knownRoomKey())
	got := hex.EncodeToString(bk[:])
	const want = "a38aaae51f23179ce85e388f8f7ade26c44593fdc8a24f59172251d0f2520104"
	if got != want {
		t.Fatalf("BeaconKey vector changed (update web interop too):\n got %s\nwant %s", got, want)
	}
}

// TestBeaconKeyDeterministicAndDistinct proves the beacon key is deterministic for
// a given room key and is NOT the room key itself — a beacon reader can never
// recover the message key from it.
func TestBeaconKeyDeterministicAndDistinct(t *testing.T) {
	rk := knownRoomKey()
	a := BeaconKey(rk)
	b := BeaconKey(rk)
	if a != b {
		t.Fatal("BeaconKey is not deterministic")
	}
	if bytes.Equal(a[:], rk.Key[:]) {
		t.Fatal("beacon key must be DISTINCT from the room/message key")
	}
	// A different room key yields a different beacon key.
	var rk2 RoomKey
	for i := range rk2.Key {
		rk2.Key[i] = byte(0xFF - i)
	}
	if BeaconKey(rk2) == a {
		t.Fatal("different room keys must yield different beacon keys")
	}
}

// TestBeaconSealOpenRoundTrip proves a sealed beacon decrypts back to the same
// status, and that a wrong key or tampered blob fails.
func TestBeaconSealOpenRoundTrip(t *testing.T) {
	bk := BeaconKey(knownRoomKey())
	const status = "Cause isolated, mitigation deploying, ETA 20m"

	blob, err := SealBeacon(bk, status)
	if err != nil {
		t.Fatalf("seal: %v", err)
	}
	got, err := OpenBeacon(bk, blob)
	if err != nil || got != status {
		t.Fatalf("round-trip: got %q err %v", got, err)
	}

	// A wrong beacon key cannot read it.
	var wrong [32]byte
	if _, err := OpenBeacon(wrong, blob); err == nil {
		t.Fatal("a wrong key must not decrypt the beacon")
	}
	// A tampered blob is rejected (AEAD integrity).
	blob[len(blob)-1] ^= 0xFF
	if _, err := OpenBeacon(bk, blob); err == nil {
		t.Fatal("a tampered blob must not decrypt")
	}
}

// TestGenBeaconInteropVector prints a Go-produced beacon vector — the derived
// beacon key and a fixed-nonce ciphertext blob — for the web reader's interop test
// to decrypt, proving the browser decrypts exactly what the TUI encrypts. Skipped
// in normal runs:
//
//	GEN_INTEROP=1 go test ./tui/internal/crypto -run TestGenBeaconInteropVector -v
func TestGenBeaconInteropVector(t *testing.T) {
	if os.Getenv("GEN_INTEROP") == "" {
		t.Skip("set GEN_INTEROP=1 to print the beacon interop vector")
	}
	bk := BeaconKey(knownRoomKey())
	const status = "SEV1 declared, cause under investigation"

	// Fixed nonce so the blob is reproducible for the pinned web vector.
	var nonce [chacha20poly1305.NonceSizeX]byte
	for i := range nonce {
		nonce[i] = byte(i)
	}
	aead, _ := chacha20poly1305.NewX(bk[:])
	ct := aead.Seal(nil, nonce[:], []byte(status), nil)
	blob := append(nonce[:], ct...)

	t.Logf("\n--- BEGIN BEACON INTEROP VECTOR ---\n"+
		"beaconKeyHex: %s\n"+
		"beaconKeyB64: %s\n"+
		"status:       %q\n"+
		"blobB64:      %s\n"+
		"--- END BEACON INTEROP VECTOR ---",
		hex.EncodeToString(bk[:]), base64.StdEncoding.EncodeToString(bk[:]),
		status, base64.StdEncoding.EncodeToString(blob))
}

File: web/src/crypto/beacon.test.ts
import { describe, it, expect } from "vitest";
import { fromB64, toB64 } from "./base64";
import { deriveBeaconKey, openBeacon } from "./beacon";

// Status Beacon interop (§1.2). Vector from crypto's TestGenBeaconInteropVector
// (room key 0x00..0x1f). Proves the browser reader decrypts exactly what the Go TUI
// sealed, using the same beacon key. (In restricted environments run the worker-free
// `npm run test:node` instead of vitest.)
const BEACON = {
  beaconKeyB64: "o4qq5R8jF5zoXjiPj3reJsRFk/3Iok9ZFyJR0PJSAQQ=",
  status: "SEV1 declared, cause under investigation",
  blobB64:
    "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXS7PWgytZDWAVsDrYMITE9OUrOPwZG3lWI/MvHt45gZZR5pZiyVYm/8+S9ocwrYnsP8ONSBOz3IE=",
};

function knownRoomKey(): Uint8Array {
  const rk = new Uint8Array(32);
  for (let i = 0; i < 32; i++) rk[i] = i;
  return rk;
}

describe("status beacon interop", () => {
  it("derives the same beacon key as Go", () => {
    expect(toB64(deriveBeaconKey(knownRoomKey()))).toBe(BEACON.beaconKeyB64);
  });

  it("decrypts a Go-sealed beacon with the link key", () => {
    expect(openBeacon(fromB64(BEACON.beaconKeyB64), fromB64(BEACON.blobB64))).toBe(BEACON.status);
  });

  it("the message room key cannot read the beacon (distinct keys)", () => {
    expect(() => openBeacon(knownRoomKey(), fromB64(BEACON.blobB64))).toThrow();
  });

  it("rejects a tampered blob", () => {
    const blob = fromB64(BEACON.blobB64);
    blob[blob.length - 1] ^= 0xff;
    expect(() => openBeacon(fromB64(BEACON.beaconKeyB64), blob)).toThrow();
  });
});
